### PR TITLE
[AIRFLOW-2333] Add Segment Hook with unit tests

### PR DESCRIPTION
**Pull request by jzucker2**
Make sure you have checked _all_ steps below.

### JIRA
- [ ] My PR addresses the following [Airflow JIRA](https://issues.apache.org/jira/browse/AIRFLOW/) issues and references them in the PR title. For example, "\[AIRFLOW-XXX\] My Airflow PR"
    - https://issues.apache.org/jira/browse/AIRFLOW-2333
    - In case you are fixing a typo in the documentation you can prepend your commit with \[AIRFLOW-XXX\], code changes always need a JIRA issue.

### Description
- [ ] Here are some details about my PR, including screenshots of any UI changes:
Add Segment Hook to store Segment Analytics access
token with encryption in metadata database.

### Tests
- [ ] My PR adds the following unit tests __OR__ does not need testing for this extremely good reason:
Unit test added.

### Commits
- [ ] My commits all reference JIRA issues in their subject lines, and I have squashed multiple commits if they address the same issue. In addition, my commits follow the guidelines from "[How to write a good git commit message](http://chris.beams.io/posts/git-commit/)":
    1. Subject is separated from body by a blank line
    2. Subject is limited to 50 characters
    3. Subject does not end with a period
    4. Subject uses the imperative mood ("add", not "adding")
    5. Body wraps at 72 characters
    6. Body explains "what" and "why", not "how"

### Documentation
- [ ] In case of new functionality, my PR adds documentation that describes how to use it.
    - When adding new operators/hooks/sensors, the autoclass documentation generation needs to be added.

### Code Quality
- [ ] Passes `git diff upstream/master -u -- "*.py" | flake8 --diff`

### CODE REVIEW

File: airflow/contrib/hooks/segment_hook.py [hunk 1]
```diff
@@ -0,0 +1,92 @@
...
+        self.connection = self.get_connection(self.segment_conn_id)
+        self.extras = self.connection.extra_dejson
+        self.write_key = self.extras.get('write_key')
+        if self.write_key is None:
+            raise AirflowException('No Segment write key provided')
+
+    def get_conn(self):
+        self.log.info('Setting write key for Segment analytics connection')
+        analytics.debug = self.segment_debug_mode
+        if self.segment_debug_mode:
```
Comment (Fokko) [inline airflow/contrib/hooks/segment_hook.py:79]: Please remove this if, if you are logging in debug mode, then you should just show everything.

File: airflow/contrib/hooks/segment_hook.py [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+        Airflow operators to move the data around or interact with segment.
+
+        :param segment_conn_id: the name of the connection that has the parameters
+                            we need to connect to Segment.
+                            The connection should be type `json` and include a
+                            write_key security token in the `Extras` field.
+        :type segment_conn_id: str
+        :param segment_debug_mode: Determines whether Segment should run in debug mode.
+        Defaults to False
+        :type segment_debug_mode: False
```
Comment (Fokko) [inline airflow/contrib/hooks/segment_hook.py:57]: Type is boolean I guess.

**Review by Fokko — CHANGES_REQUESTED** [review]
Thanks @jzucker2 for the contribution.

I looked into Segment and really like it. But contributing only a hook, doesn't really make sense to me. The hooks represent connections to the backend (in this case Segment). The hook comes with an operator, which is used by the developer to interact with. Sometimes, the more advanced Airflow users, also use hooks on its own. Could you explain what you are trying to achieve?

**Comment by tedmiston** [comment]
@jzucker2 Are you open to adding a `SendTrackEventOperator` that takes a name and a dict to pass through to `analytics.track(...)`?

I'd be curious to know if you find other event types useful here as well, maybe `SendIdentifyEventOperator` would be nice?

**Comment by r39132** [comment]
@jzucker2 please respond to comments and rebase.

**Comment by jzucker2** [comment]
@tedmiston I am definitely interested in adding the `SendTrackEventOperator`

@r39132 sorry for the delay. Been busy with some work emergencies, should have time to give this another pass with the suggested operators from @tedmiston by end of next week.

**Comment by tedmiston** [comment]
Sounds great.  Happy to help test this as well.

**Comment by r39132** [comment]
@jzucker2 ping.. just a friendly reminder.

**Comment by r39132** [comment]
Closing this for now. Please re-open when ready

**Comment by jzucker2** [comment]
@r39132 woah! I didn't even get a work day to respond. I was actually working on rebasing while you closed this.

**Comment by jzucker2** [comment]
@r39132 I added an operator and got all unit tests passing but I'm waiting on you to reopen this pull request.

**Comment by tedmiston** [comment]
I see some PRs like https://github.com/apache/incubator-airflow/pull/3197 https://github.com/apache/incubator-airflow/pull/3319 preface their title with [WIP] to make it clear to maintainers that it's still being actively developed.  Not sure if that's helpful or if we've got a standard here.

I've also seen other repos with a bot to close open PRs after e.g., 30 days of no activity.

**Comment by jzucker2** [comment]
@tedmiston I am done with this and want to merge it but I see no way to reopen this pull request in the UI. I appreciate the discussion of [WIP] but my last comment clearly indicates I am not WIP but in fact done.

As far as inactivity, this was inactive for less than half the 30 days discussed.

**Comment by r39132** [comment]
@jzucker2 Open a new PR and reference this if there is a need to keep the history.

**Pull request closed without merging**